### DISCUSSION TITLE
Using provider

### DISCUSSION THREAD

**Discussion by ArtinKennedy** [schwabe/ics-openvpn#1673] [Q&A] [answered]
Hi;
I wanna add a provider to openssl. This provider is oqs-provider. How can I set -DMODULESDIR in openssl.cmake to fine provider module correctly?

**Comment by schwabe** [comment]
Just setting MODULESDIR is probably not enough. Modern Android platforms do not allow execution of code that is not originally part of the app that easily. But adding a define in cmake is quite trivial. See https://cmake.org/cmake/help/latest/command/add_compile_definitions.html

If you explain a bit more what provider you are intending to use and why, I might be able to give more insight.

**Reply by ArtinKennedy** [reply]
Thanks schwabe.

I wanna add oqs-provider module to openssl for using liboqs.
https://github.com/open-quantum-safe/oqs-provider.git
https://github.com/open-quantum-safe/liboqs.git

I set MODULESDIR=ossl-modules, and in runtime confront with this error: dlopen can not find ossl-modules/oqsprovider.so.
And I have no idea for resolving this problem.

**Reply by schwabe** [reply, marked as answer]
I am not sure where you are setting the MODULESDIR. But you might need to have the absolute path to be able to load the provider.  

As for building oqs-provider and liboqs on Android itself. That might be a bit more more involved if nobody has done that before. I am not entirely sure how well current Android versions like dynamically loaded libraries. But qos-provider also supports being used as internal provider (https://github.com/open-quantum-safe/oqs-provider/blob/main/CONFIGURE.md#oqs_provider_build_static).
